core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
 
   @Override
   public void perform() {
-    if (!connected) {
-      return; // Don't waste any time converting images if there's no dashboard connected
+    final long now = System.nanoTime(); // NOPMD
+
+    if (!cscoreLoaded) {
+      throw new IllegalStateException(
+          "cscore could not be loaded. The image streaming server cannot be started.");
     }
 
-    if (inputSocket.getValue().get().empty()) {
+    opencv_core.Mat input = inputSocket.getValue().get();
+    if (input.empty() || input.isNull()) {
       throw new IllegalArgumentException("Input image must not be empty");
     }
 
-    synchronized (imageLock) {
-      imencode(".jpeg", inputSocket.getValue().get(), imagePointer,
-          new IntPointer(CV_IMWRITE_JPEG_QUALITY, qualitySocket.getValue().get().intValue()));
-      imageLock.notifyAll();
+    copyJavaCvToOpenCvMat(input, publishMat);
+    // Make sure the output resolution is up to date. Might not be needed, depends on cscore updates
+    serverSource.setResolution(input.size().width(), input.size().height());
+    serverSource.putFrame(publishMat);
+    if (lastFrame != -1) {
+      long dt = now - lastFrame;
+      serverSource.setFPS((int) (1e9 / dt));
     }
+    lastFrame = now;
   }
 
   @Override
   public synchronized void cleanUp() {
-    // Stop the video server if there are no Publish Video steps left
-    serverThread.interrupt();
     numSteps--;
+    if (cscoreLoaded) {
+      availablePorts.addFirst(server.getPort());
+      ourTable.getKeys().forEach(ourTable::delete);
+      serverSource.setConnected(false);
+      serverSource.free();
+      server.free();
+    }
+  }
+
+  /**
+   * Copies the data from a JavaCV Mat wrapper object into an OpenCV Mat wrapper object so it's
+   * usable by the {@link CvSource} for this operation.
+   *
+   * <p>Since the JavaCV and OpenCV bindings both target the same native version of OpenCV, this is
+   * implemented by simply changing the OpenCV Mat's native pointer to be the same as the one for
+   * the JavaCV Mat. This prevents memory copies and resizing/reallocating memory for the OpenCV
+   * wrapper to fit the source image. Updating the pointer is a simple field write (albeit via
+   * reflection), which is much faster and easier than allocating and copying byte buffers.</p>
+   *
+   * <p>A caveat to this approach is that the memory layout used by the OpenCV binaries bundled with
+   * both wrapper libraries <i>must</i> be identical. Using the same OpenCV version for both
+   * libraries should be enough.</p>
+   *
+   * @param javaCvMat the JavaCV Mat wrapper object to copy from
+   * @param openCvMat the OpenCV Mat wrapper object to copy into
+   * @throws RuntimeException if the OpenCV native pointer could not be set
+   */
+  private static void copyJavaCvToOpenCvMat(opencv_core.Mat javaCvMat, Mat openCvMat)
```

Comment:
Can you add some unit tests for this function to make sure that it gets covered. I don't want someone in the future to be updating dependencies and have this break because this code doesn't get run by a test.